### ISSUE TITLE
Making clearer what namespaces are public by use of underscores

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#14360]
When we add new `.py` files, we are careful to add underscores to the file names to ensure everyone sees at a glance that this is not a new public namespace. However, we have lots of files that are missing that underscore. This is only because of historical reasons; in the early days of SciPy no one paid attention to this. We had a discussion about it once, probably about 9-10 years ago, and decided to not clean things up back then because the concrete benefits were unclear.

The public namespaces are documented in http://scipy.github.io/devdocs/reference/index.html#api-definition. No one reads the docs though. This has become much more of a problem recently, because authors of other libraries have started to reimplement parts of the SciPy API, for example:
- CuPy, see https://docs.cupy.dev/en/stable/reference/scipy.html
- JAX, see https://jax.readthedocs.io/en/latest/jax.scipy.html
- `dask_image`, see http://image.dask.org/en/latest/api.html
- PyTorch is adding a `torch.special` namespace that mirrors `scipy.special`: https://pytorch.org/docs/master/special.html

Those libraries can get it wrong by accident, or they just mirror what our actual filenames are even though we tell them it's not meant to be a public namespace. An example of each:
- https://github.com/scipy/scipy/issues/10204 discusses `dask_image` restructuring its namespace to match `scipy.ndimage`
- https://github.com/cupy/cupy/pull/5047 is about `scipy.stats.distributions`

Long story short: we should clean this up. There's two ways we can go about this: 
1. don't deprecate anything. Just move `somefile.py` to `_somefile.py`, and then add a new `somefile.py` which reimports the public functions from `_somefile.py` and adds a big comment with warnings about this not being a public namespace at the top of the file.
2. do (1), and then deprecate the functions in `somefile`

I'd suggest to do (2) by default, and only make exceptions in case something is quite heavily used and adding deprecations would be too disruptive.

We should prioritize modules that are being duplicated in CuPy, JAX, PyTorch et al.

EDIT: a test should also be added that no new private-but-public-looking namespaces are added. The approach can be taken from https://github.com/numpy/numpy/blob/main/numpy/tests/test_public_api.py

EDIT 2: a basic API design principle is: _a public object should only be available from one namespace_. Having any function in two or more places is just extra technical debt, and with things like dispatching on an API or another library implementing a mirror API, the cost goes up.
```python
>>> from scipy import ndimage
>>> ndimage.filters.gaussian_filter is ndimage.gaussian_filter  # :(
True
```

EDIT 3: now that it's decided we're going ahead with this, here is a tracker of which modules are done:

- [x] cluster
- [x] constants
- [x] fft
- [x] fftpack
- [x] integrate
- [x] interpolate
- [x] io
- [x] linalg
- [x] misc
- [x] ndimage
- [x] odr
- [x] optimize
- [x] signal
- [x] sparse
- [x] sparse.csgraph
- [x] sparse.linalg
- [x] spatial
- [x] special
- [x] stats

Also, we should take over the public API test from NumPy:

- [x]  https://github.com/numpy/numpy/blob/main/numpy/tests/test_public_api.py

EDIT 4: 

- [x] right before closing this issue, we should compare master with 1.7.0 and use some version of `sorted([x for x in dir(fitpack) if not (x.startswith('_') or x in ('np', 'warnings'))]` to compare for each file/namespace we gave an underscore that everything that looks like a regular SciPy function/object is present in the listings from which we raise deprecation warnings. That way, even if people used clearly private things that were not in `__all__` in that file, they will still get the warnings rather than a hard break.

**Comment by rgommers** [comment]
To check the impact on git history I pushed a branch here: https://github.com/rgommers/scipy/blame/test-renaming-file/scipy/ndimage/. It has 3 new commits:
1. rename `ndimage/filters.py` to `_filters.py`
2. add a new file `filters.py`
3. make a random change in `_filters.py`

`git blame` is not affected at all, it does not show the rename on the command line nor in the GitHub UI:
```bash
$ git blame -L30,+5 _filters.py
ca465a651f6 Lib/ndimage/Lib/filters.py (Ed Schofield     2006-03-18 13:52:58 +0000 30) 
7e3cda70f54 scipy/ndimage/filters.py   (Gregory R. Lee   2020-02-01 14:41:31 -0700 31) from collections.abc import Iterable
9ccbb172e6a scipy/ndimage/filters.py   (Aditya Bharti    2018-03-12 03:14:07 +0530 32) import warnings
ee4db9c301d Lib/ndimage/filters.py     (Robert Kern      2006-09-24 09:05:13 +0000 33) import numpy
69cfe48f37f scipy/ndimage/filters.py   (Anju Geetha Nair 2018-11-04 01:14:40 -0800 34) import operator
```

<img width="665" alt="image" src="https://user-images.githubusercontent.com/98330/124647316-155fe980-de96-11eb-8b6e-48c3052d85bc.png">

If you check the history of the file, you do see a change:  

<img width="336" alt="image" src="https://user-images.githubusercontent.com/98330/124647385-2dd00400-de96-11eb-99c3-048daf49b75c.png">

That change is the main (only?) downside, but it's a minor one. Locally you can ignore that rename with `--follow`, e.g.:
```bash
$ gitk --follow _filters.py
```

**Comment by AnirudhDagar** [comment]
Thanks, @rgommers for throwing some light on the history of these private namespaces which look like public (without leading underscores). Also, thanks for verifying that renaming still keeps `git blame` intact.

I'd also lean towards this approach, but I'll wait for more discussion (if any) on this.
> do (1), and then deprecate the functions in somefile

Once we have decided, I can start with `scipy.ndimage.<module>` -> `scipy.ndimage._<module>`.

**Comment by rgommers** [comment]
Message to the mailing list about this: https://mail.python.org/pipermail/scipy-dev/2021-July/024943.html

**Comment by tylerjereddy** [comment]
I *think* I understand what you're proposing, seems to make sense. One of the drawbacks compared to aggressively private languages I suppose (Rust comes to mind--you need a `pub` in front of pretty much anything you want public).

**Comment by rgommers** [comment]
> I _think_ I understand what you're proposing, seems to make sense.

I opened a PR with an example to make sure we're on the same page: gh-14419.

**Comment by WarrenWeckesser** [comment]
These two questions on stackoverflow

* https://stackoverflow.com/questions/69001787/vs-code-python-doesnt-autocomplete-scipy-stats
* https://stackoverflow.com/questions/69774502/vscode-does-not-find-module-norm-in-scipy-stats

might be the beginning of  a surge of questions about automatic completion of scipy names in VSCode.  The comments and answers in that first question suggest that the pylance plugin of VSCode doesn't autocomplete names in the public API if they *originate* from modules that begin with an underscore.  (Note: I haven't confirmed that claim.) If that is true, then it is likely that the changes in the namespace clean up will increase the occurrence of mysterious code completion failures in VSCode, at least until pylance is fixed.

To be clear: I'm definitely in favor of the clean up. This is just advance warning about an issue with VSCode that might be exacerbated by the clean  up.

**Comment by rgommers** [comment]
Thanks for pointing that out @WarrenWeckesser! I think Pylance is indeed already not working with `from xxx import *` imports from underscored files, which we already had a lot of - we're just adding more.

There is an orthogonal question here: should we get rid of all star imports? It is technically best practice to not use star imports, we'll just get really long file listings in `__init__.py` files. But it's not hard to clean up, just tedious.

**Comment by czgdp1807** [comment]
I am taking up `optimize` next. Since, `sparse` PR is still open, once that's merged I will go with `sparse.linalg`.

**Comment by Smit-create** [comment]
I'll take `special` next

**Comment by tupui** [comment]
> There is an orthogonal question here: should we get rid of all star imports? It is technically best practice to not use star imports, we'll just get really long file listings in `__init__.py` files. But it's not hard to clean up, just tedious.

I am personally in favour of this. It's more explicit and I find it easier to find your way back this way.

**Comment by rgommers** [comment]
One more thing to do here. @Smit-create it would be great to have some kind of script that can walk the SciPy import tree for all module-like things that don't start with an underscore, and then save the result of `dir(modulename)`. So we can run it on `1.7.0` and `master`, and compare the results.

**Comment by h-vetinari** [comment]
Was it intentional that a `conftest.py` get's picked up as a public module in your PR, @Smit-create?

I mean, it's a separate debate how much (if at all) of the test suite should be parametrized & fixturized, but it seems to me that something that's testing-specific like that (and a well-documented [pattern](https://docs.pytest.org/en/6.2.x/example/simple.html) in pytest) should not be triggering the "public-module" error.

Context: #15073 passed before #15040 was merged (if the approach in my PR is found useful, I can probably add an exclusion for `conftest.py` on top of what you did in #15040; I mostly just wanted to bring this up in the issue for discussion)

**Comment by Smit-create** [comment]
> Was it intentional that a `conftest.py` get's picked up as a public module in your PR, @Smit-create?

Yes, it's private but unfortunately we cannot underscore it - see https://stackoverflow.com/questions/63236001/in-pytest-how-can-i-specify-a-different-conftest-py-file-to-use.

**Comment by h-vetinari** [comment]
> Yes, it's private but unfortunately we cannot underscore it

Yes, which is why it should be excluded from the public-API check...

**Comment by Smit-create** [comment]
> One more thing to do here. @Smit-create it would be great to have some kind of script that can walk the SciPy import tree for all module-like things that don't start with an underscore, and then save the result of `dir(modulename)`. So we can run it on `1.7.0` and `master`, and compare the results.

For comparing the results, I followed these steps:
1. Ran [check_save_api.py](https://gist.github.com/Smit-create/8233e265bb7107ddd891e8338f7fd76a) for both the versions and saved the results into two files.
2. I used these saved results to compare two APIs using this [compare.py](https://gist.github.com/Smit-create/f7d640dc1b1be2d1f16029984ff4979b)

The results I get are:

```
These files are not present in current API:
scipy.interpolate.interpnd_info
scipy.io.harwell_boeing.hb
scipy.io.matlab.byteordercodes
scipy.io.matlab.mio
scipy.io.matlab.mio4
scipy.io.matlab.mio5
scipy.io.matlab.mio5_params
scipy.io.matlab.mio5_utils
scipy.io.matlab.mio_utils
scipy.io.matlab.miobase
scipy.io.matlab.streams
scipy.sparse.generate_sparsetools
scipy.sparse.linalg.dsolve.linsolve
scipy.sparse.linalg.eigen.arpack
scipy.sparse.linalg.eigen.arpack.arpack
scipy.sparse.linalg.eigen.lobpcg
scipy.sparse.linalg.eigen.lobpcg.lobpcg
scipy.sparse.linalg.isolve.iterative
scipy.sparse.linalg.isolve.lgmres
scipy.sparse.linalg.isolve.lsmr
scipy.sparse.linalg.isolve.lsqr
scipy.sparse.linalg.isolve.minres
scipy.sparse.linalg.isolve.utils
----------------------------------
No files are breaking imports
```

**Comment by h-vetinari** [comment]
> Yes, which is why it should be excluded from the public-API check...

@Smit-create and whoever has an opinion about this: could you please review https://github.com/scipy/scipy/pull/15073/commits/e7113f8f1c789d9bc6adff916dd1c8e168c68cd3 in #15073?

**Comment by rgommers** [comment]
> could you please review `e7113f8` in #15073?

I'd recommend to not mix that in with an unrelated `scipy.spatial` PR. Also, why does this matter at all? We have only a single `conftest.py` in the source tree, so your version just looks less readable while doing the same thing.

**Comment by h-vetinari** [comment]
> I'd recommend to not mix that in with an unrelated `scipy.spatial` PR.

I can break it out into a separate PR, but since there was no reaction (other than a 👍) here, I felt like I need to give it a shot myself.

> Also, why does this matter at all? We have only a single `conftest.py` in the source tree

Well, this is what #15073 changes. And that's not a crazy fancy that I came up with, but it's a core design aspect of using `conftest.py`, that they can live on every level of the test suite hierarchy, and only get applied to the tests which lie below. In this case, having a `scipy.spatial`-specific fixture should reasonably live in `scipy/spatial/conftest.py` - otherwise the global `conftest.py` gets huge, full of overlaps, and eventually unmaintainable.

I mean, the placement of the `conftest.py` (or whether to use a fixture at all) is not something I'm deeply invested in in the context of #15073 - if there's strong opposition, I'm happy to drop it. I just started with a best-practice approach (IMO).

**Comment by rgommers** [comment]
Ah okay, that does sound reasonable - thanks for the explanation, let's leave that as is in your PR then @h-vetinari.

**Comment by rgommers** [comment]
I see a few more @Smit-create:
```
scipy.integrate.odepack
scipy.integrate.quadpack
scipy.interpolate.interpnd_info
scipy.io.harwell_boeing.hb
scipy.signal.sigtools
scipy.sparse.generate_sparsetools
scipy.sparse.linalg.dsolve.linsolve
scipy.sparse.linalg.eigen.arpack
scipy.sparse.linalg.eigen.arpack.arpack
scipy.sparse.linalg.eigen.lobpcg
scipy.sparse.linalg.eigen.lobpcg.lobpcg
scipy.sparse.linalg.isolve.iterative
scipy.sparse.linalg.isolve.lgmres
scipy.sparse.linalg.isolve.lsmr
scipy.sparse.linalg.isolve.lsqr
scipy.sparse.linalg.isolve.minres
scipy.sparse.linalg.isolve.utils
scipy.sparse.spfuncs
```
The ones that are easy to address by adding back a `.py` file which raises deprecation warnings are:
```
scipy.integrate.odepack
scipy.integrate.quadpack
scipy.signal.sigtools
scipy.sparse.spfuncs
```
For the others:
- `interpolate.interpnd_info` has no functions in it without underscores, so no need to add anything back
- `generate_sparsetools` is a utility script that we shouldn't even install, so doesn't need to be dealt with
- `io.harwell_boeing.hb` and all the `sparse.linalg` ones cannot be added back because the whole directory is gone. This should be fine.

Can you add `.py` files for the four that can be dealt with @Smit-create?

**Comment by rgommers** [comment]
For the second part, the `dir(modname)` bit is wrong @Smit-create - if you inspect the `cur_api.txt` file you'll see that what is stored there now is:
```
{'scipy.cluster': ['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isascii', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'removeprefix', 'removesuffix', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']
```
So what's going on there is that `modname` is a string rather than a module object, and hence `dir` gives all string methods. Can you try to fix that up? I expect there'll be a lot of things missing in `__all__` lists, because for the initial PRs we filled them with public functions rather than with `dir(modname)` output.

**Comment by Smit-create** [comment]
> For the second part, the `dir(modname)` bit is wrong @Smit-create

Yes, you are right.

I updated the files [check_save_api.py](https://gist.github.com/Smit-create/8233e265bb7107ddd891e8338f7fd76a) and [compare.py](https://gist.github.com/Smit-create/f7d640dc1b1be2d1f16029984ff4979b)

Again, I followed these steps:

1. Ran [check_save_api.py](https://gist.github.com/Smit-create/8233e265bb7107ddd891e8338f7fd76a) for both the versions and saved the results into two files.
2. I used these saved results to compare two APIs using this [compare.py](https://gist.github.com/Smit-create/f7d640dc1b1be2d1f16029984ff4979b)

The results I get are: [result](https://gist.github.com/Smit-create/faf7eaffa642fc737258a742affce219)

**Comment by rgommers** [comment]
That output looks good to me now, thanks @Smit-create.

**Comment by rgommers** [comment]
The last fixes went in in gh-15105, this is all done now so I'll close the issue. Thanks @Smit-create, @czgdp1807, and everyone else who helped with this!

**Comment by h-vetinari** [comment]
Just noticed while [testing](https://github.com/conda-forge/scipy-feedstock/pull/198) that `scipy.signal.sigtools` now is missing without a deprecation warning. This might be intentional, and I'm fine with it, but given that they were mentioned [above](https://github.com/scipy/scipy/issues/14360#issuecomment-977161594) by @rgommers as being supposed to warn, I thought I'd mention it.

**Comment by Smit-create** [comment]
Hi @h-vetinari, I left it intentionally because it only had private functions. See https://github.com/scipy/scipy/pull/15105#issuecomment-980842147

**Comment by h-vetinari** [comment]
> Hi @h-vetinari, I left it intentionally because it only had private functions. See [#15105 (comment)](https://github.com/scipy/scipy/pull/15105#issuecomment-980842147)

Fine by me, just wanted to make sure it was intentional and not an oversight.

**Comment by h-vetinari** [comment]
I re-read this issue and was wondering when removal was envisaged. Turns out it's mentioned in the mailing list [post](https://mail.python.org/archives/list/scipy-dev@python.org/message/I4B27XJY6RZVWNXF6F62Q7KX62WXKBQF/) that the removal is currently planned for scipy 2.0. I'll reflect it as such in the deprecation tracker.

**Comment by mdhaber** [comment]
Were sub-sub-packages supposed to be included here? There are several that still expose their innards.

```python3
from scipy.cluster.vq import xpx, np, check_random_state
from scipy.cluster.hierarchy import xp_capabilities, array_namespace, is_dask
from scipy.io.wavfile import warnings, struct, sys
```

**Comment by j-bowhay** [comment]
> Were sub-sub-packages supposed to be included here? There are several that still expose their innards.
> 
> from scipy.cluster.vq import xpx, np, check_random_state
> from scipy.cluster.hierarchy import xp_capabilities, array_namespace, is_dask
> from scipy.io.wavfile import warnings, struct, sys
> from scipy.linalg.blas import functools
> from scipy.linalg.lapack import np
> from scipy.linalg.interpolative import np
> from scipy.spatial.distance import np
> from scipy.stats.contingency import np

I noticed this the other day too when you posted prototype array api table. IMO it would be good to clean this up.

**Comment by rgommers** [comment]
Yes agreed, it should apply to all public API. Please don't modify years-old closed issues and remove the milestone though. I'll open a new issue for a follow-up. It should also be investigated if these subsubmodules are missing from the public API testing.

**Comment by rgommers** [comment]
Follow-up issue: gh-23033
